protocol/consistencychecks.go

```diff
@@ -0,0 +1,307 @@
+// when she registers her binding with a ConiksDirectory.
+// This ConsistencyChecks instance then will be used to verify
+// the returned responses from the ConiksDirectory.
+type ConsistencyChecks struct {
+	SavedSTR *m.SignedTreeRoot
+
+	// extensions settings
+	useTBs bool
+	TBs    map[string]*TemporaryBinding
+
+	// signing key
+	signKey sign.PublicKey
+}
+
+// NewCC creates an instance of ConsistencyChecks using
+// the pinned directory's STR at epoch 0 or
+// the consistency state read from a persistent storage.
+func NewCC(savedSTR *m.SignedTreeRoot, useTBs bool, signKey sign.PublicKey) *ConsistencyChecks {
+	// TODO: see #110
+	if !useTBs {
+		panic("[coniks] Currently the server is forced to use TBs.")
+	}
+	cc := &ConsistencyChecks{
+		SavedSTR: savedSTR,
+		useTBs:   useTBs,
+		signKey:  signKey,
+	}
+	if useTBs {
+		cc.TBs = make(map[string]*TemporaryBinding)
+	}
+	return cc
+}
+
+// HandleResponse verifies the directory's response for a request.
+// It first verifies the directory's returned status code of the request.
+// If the status code is not in the ErrorResponses array, it means
+// the directory has successfully handled the request.
+// The verifier will then verify the consistency state of the response.
+// This will panic if it is called with an int
+// which isn't a valid/known request-type.
+// Note that the consistency state would be updated regardless of
+// whether the checks pass / fail, since it contains proof of being issued.
+func (cc *ConsistencyChecks) HandleResponse(requestType int, msg *Response,
+	uname string, key []byte) ErrorCode {
+	if ErrorResponses[msg.Error] {
+		return msg.Error
+	}
+	switch requestType {
+	case RegistrationType, KeyLookupType:
+		if _, ok := msg.DirectoryResponse.(*DirectoryProof); !ok {
+			return ErrorMalformedDirectoryMessage
+		}
+	default:
+		panic("[coniks] Unknown request type.")
+	}
+	if err := cc.updateSTR(requestType, msg); err != nil {
+		return err.(ErrorCode)
+	}
+	if err := cc.checkConsistency(requestType, msg, uname, key); err != Passed {
+		return err
+	}
+	if err := cc.updateTBs(requestType, msg, uname, key); err != nil {
+		return err.(ErrorCode)
+	}
+	return Passed
+}
+
+func (cc *ConsistencyChecks) updateSTR(requestType int, msg *Response) error {
+	var str *m.SignedTreeRoot
+	switch requestType {
+	case RegistrationType, KeyLookupType:
+		str = msg.DirectoryResponse.(*DirectoryProof).STR
+		// First response
+		if cc.SavedSTR == nil {
+			cc.SavedSTR = str
+			return nil
+		}
+		// FIXME: check whether the STR was issued on time and whatnot.
+		// Maybe it has something to do w/ #81 and client transitioning between epochs.
+		// Try to verify w/ what's been saved
+		if err := cc.verifySTR(str); err == nil {
+			return nil
+		}
+		// Otherwise, expect that we've entered a new epoch
+		if err := cc.verifySTRConsistency(cc.SavedSTR, str); err != nil {
+			return err
+		}
+	}
+
+	// And update the saved STR
+	cc.SavedSTR = str
```

Comment:
If we had any other request types, this would overwrite with `nil`, right.

Can we all a `default` case that panics?